analyzer/src/main/kotlin/managers/SpdxTransitiveDocument.kt

```diff
@@ -0,0 +1,329 @@
+import org.ossreviewtoolkit.model.OrtIssue
+import org.ossreviewtoolkit.model.Severity
+import org.ossreviewtoolkit.model.createAndLogIssue
+import org.ossreviewtoolkit.utils.common.safeDeleteRecursively
+import org.ossreviewtoolkit.utils.core.OkHttpClientHelper
+import org.ossreviewtoolkit.utils.core.addBasicAuthorization
+import org.ossreviewtoolkit.utils.core.createOrtTempDir
+import org.ossreviewtoolkit.utils.core.downloadFile
+import org.ossreviewtoolkit.utils.core.log
+import org.ossreviewtoolkit.utils.spdx.model.SpdxDocument
+import org.ossreviewtoolkit.utils.spdx.model.SpdxExternalDocumentReference
+import org.ossreviewtoolkit.utils.spdx.model.SpdxPackage
+
+/**
+ * A data class storing information about a root SPDX document and all the documents referenced by it.
+ *
+ * This class is used by [SpdxDocumentFile] to get a combined view on all packages and their relations defined in a set
+ * of SPDX files.
+ */
+internal data class SpdxTransitiveDocument(
+    /**
+     * The root document. This is the starting point, from which all external references have been traversed.
+     */
+    val rootDocument: SpdxDocument,
+
+    /**
+     * The name of the package manager that uses this document. This is mainly used to fill the source in generated
+     * [OrtIssue]s.
+     */
+    val managerName: String,
+
+    /**
+     * Holds a map with all [SpdxDocument]s that are referenced directly or indirectly from the root document, using
+     * the external reference objects as keys.
+     */
+    val referencedDocuments: Map<SpdxExternalDocumentReference, SpdxDocument>,
+
+    /**
+     * A map allowing direct access to the packages declared in one of the contained documents. For the packages of the
+     * root document, the key is the package's identifiers. For other packages, the identifier needs to be prefixed
+     * with the document identifier of the reference.
+     */
+    private val packagesById: Map<String, SpdxPackage>,
+
+    /**
+     * A map storing issues that were encountered when resolving external document references. These issues are also
+     * assigned to [SpdxPackage]s defined in the corresponding external documents.
+     */
+    private val issuesByReferenceId: Map<String, OrtIssue>
+) {
+    companion object {
+        fun load(loader: SpdxDocumentLoader, rootDocumentFile: File, managerName: String): SpdxTransitiveDocument {
+            val rootDocument = loader.load(rootDocumentFile)
+
+            val references = mutableMapOf<SpdxExternalDocumentReference, SpdxDocument>()
+            val issues = mutableMapOf<String, OrtIssue>()
+            resolveTransitiveReferences(
+                loader,
+                managerName,
+                rootDocument,
+                rootDocumentFile.toURI(),
+                references,
+                issues,
+                mutableSetOf()
+            )
+
+            val packages = collectPackages(references) + rootDocument.getPackages()
```

Comment:
The `+` potentially overwrites entries in the map. Is this intended? Should we use our own `zip(WithDefault)` functions instead? If there never could be a key-clash, please add a code comment why.

Reply:
Added a comment.

Reply:
BTW, I was in particular thinking about cyclic dependencies. Would this be that place to catch them?

Reply:
No, cycles in the dependencies are handled by `SpdxDocumentFile` when processing the relationships. `SpdxResolvedDocument` detects cycles in references in the `resolveAllReferences()` function.